[TTAHUB-1254] Validate URLS on the backend and before saving drafts

thewatermethod opened the pull request:
## Description of change
On save draft, bad URLs could be sent to the backend. This change
- Prevents bad urls from being sent
- And if they are, we now make sure to validate them on the backend (no invalid URLs are saved)

## How to test
To duplicate bug - save draft with invalid URL in the resource field of the AR or RTR.
To test fix, enter invalid URL text in an AR or the RTR form. Attempt to save or save draft. Note that your invalid text is not saved.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1254


## Checklists

### Every PR

<!-- Add details to each completed item -->
- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [n/a] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [n/a] API Documentation updated
- [n/a] Boundary diagram updated
- [n/a] Logical Data Model updated
- [n/a] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

thewatermethod commented on src/lib/urlUtils.js:
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable import/prefer-default-export */
+// regex to match a valid url, it must start with http:// or https://, have at least one dot, and not end with a dot or a space
+const VALID_URL_REGEX = /^https?:\/\/.*\.[^ |^.]/;
+
```
sharing code between the front and the backend continues to be a pain point

thewatermethod commented on src/services/createOrUpdateGoalsForActivityReport.test.js:
```diff
@@ -353,7 +353,7 @@ describe('createOrUpdateGoalsForActivityReport', () => {
     expect(createdGoals[0].name).toBe('Test create goal for activity reports');
-    expect(createdGoals[0].grantIds).toStrictEqual(grants.map((g) => g.id));
+    expect(createdGoals[0].grantIds.sort()).toStrictEqual(grants.map((g) => g.id).sort());
     expect(createdGoals[0].objectives.length).toBe(3);
```
this is just for deflaking reasons, nothing to do with the larger change

GarrettEHill commented on src/lib/urlUtils.js:
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable import/prefer-default-export */
+// regex to match a valid url, it must start with http:// or https://, have at least one dot, and not end with a dot or a space
+const VALID_URL_REGEX = /^https?:\/\/.*\.[^ |^.]/;
```
This is the regex i needed to cleanly pull urls out of the legacy data.
```suggestion
const VALID_URL_REGEX = '(?:(?:http|ftp|https|file):\\/\\/)?(?:[\\w%_-]+(?:(?:\\.[\\w%_-]+)+)|(?:\\/[\\w][:]))(?:[\\w\\\\\'\'.,@?^=%&:\\/~+#()-]*[\\w@?^=%&\\/~+#-])';
```

GarrettEHill commented on frontend/src/components/GoalForm/constants.js:
```diff
@@ -4,7 +4,7 @@ import { v4 as uuidv4 } from 'uuid';
 // regex to match a valid url, it must start with http:// or https://, have at least one dot, and not end with a dot or a space
 const VALID_URL_REGEX = /^https?:\/\/.*\.[^ |^.]/;
```
```suggestion
const VALID_URL_REGEX = '(?:(?:http|ftp|https|file):\\/\\/)?(?:[\\w%_-]+(?:(?:\\.[\\w%_-]+)+)|(?:\\/[\\w][:]))(?:[\\w\\\\\'\'.,@?^=%&:\\/~+#()-]*[\\w@?^=%&\\/~+#-])';
```

thewatermethod replied:
We don't want to allow file or ftp. This also looks a little complicated compared to what we actually need according to Angela, which is written out in the comment above. Can you explain why we should make this change?

thewatermethod replied:
tried adding this locally, seems to have some syntax errors with the slashes in there so I'm going to hold off committing this suggestion for now

kryswisnaskas reviewed (commented):
Works as far as not saving multiple url entries in one field!

Shouldn't we mention that these are going to be discarded?

After all, if not carefully read, our tooltip could be interpreted as inviting users to paste multiple urls in the "Link" (singular) field.
![image](https://user-images.githubusercontent.com/28231976/205675354-a4d4a692-0119-4425-be56-edd501ad8583.png)

nvms commented:
> ...if not carefully read, our tooltip could be interpreted as inviting users to paste multiple urls in the "Link" (singular) field.

piggybacking on this, I could see something like the following occurring, which passes validation. not sure about the first with the \n and \r, but the second one seems plausible

<img width="533" alt="Screen Shot 2022-12-05 at 8 33 49 AM" src="https://user-images.githubusercontent.com/17074357/205677573-9707d04e-06c0-46a0-9e40-f5450969a7e4.png">

thewatermethod commented:
@nvms 
> to be fair I can't actually reproduce that behavior of pasting from somewhere WITHOUT it sticking spaces between the urls - but just thought I'd point it out. maybe not something to worry about?

Initially, I got everyone to agree that if it was a valid URL in the loosest sense (what passes an input type="url"), then we'd be good to go. However, it's apparent that this is just not good enough in practice, so I've added a check for multiple occurrences of "http", among the other things it needs to check for

@kryswisnaskas 
> Shouldn't we mention that these are going to be discarded?

Kelly, Adam, and I updated the messaging. There are expanded criteria enumerated in the error message, and there is no helper text explicitly asking the users to enter one per line. We don't mention the discarding but hopefully this will help clarify. If there is still confusion/reaction to this, I will ask the UX team to revisit

AdamAdHocTeam commented on package.json:
```diff
@@ -15,9 +15,9 @@
-    "server": "nodemon src/index.js --exec babel-node --ignore ./tests/ --ignore ./playwright-report/ --ignore ./test-results/",
-    "worker": "nodemon src/worker.js --exec babel-node --ignore ./tests/ --ignore ./playwright-report/ --ignore ./test-results/",
-    "server:debug": "nodemon --inspect=0.0.0.0:9229 src/index.js --exec babel-node --ignore ./tests/ --ignore ./playwright-report/ --ignore ./test-results/",
+    "server": "nodemon src/index.js --exec babel-node --ignore ./tests/ --ignore ./playwright-report/ --ignore ./test-results/ --ignore ./frontend/",
```
Is this supposed to be here?

nvms replied:
whoops! PR to fix this is here: https://github.com/HHS/Head-Start-TTADP/pull/1220

AdamAdHocTeam reviewed (commented):
One minor thing I noticed is that if a bad URL (resource) is entered on the RTR. You can 'Save draft' and it will automatically remove the bad resource. However 'Save and continue' prevents the save unlike from the AR. Were the bad resource is removed.

thewatermethod commented:
> One minor thing I noticed is that if a bad URL (resource) is entered on the RTR. You can 'Save draft' and it will automatically remove the bad resource. However 'Save and continue' prevents the save unlike from the AR. Were the bad resource is removed.

@AdamAdHocTeam -> I have adjusted the behaviors to match

kryswisnaskas commented:
> @AdamAdHocTeam -> I have adjusted the behaviors to match
Would it be hard to extend the same behavior to auto save as well?

thewatermethod commented:
> Would it be hard to extend the same behavior to auto save as well?

In what way? Prevent auto save?

kryswisnaskas commented:
> * https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1254



> > Would it be hard to extend the same behavior to auto save as well?
> 
> In what way? Prevent auto save?

Flagging an invalid entry and keeping the values. The concern is that we would wipe out the potentially carefully entered urls on auto save.

thewatermethod commented:
> Flagging an invalid entry and keeping the values. The concern is that we would wipe out the potentially carefully entered urls on auto save. 

Ah, I see. Right now we do the opposite in all cases (RTR & AR). I am looking into the autosave behavior and will try to come up with a way to preserve the data at least until they navigate away
